Fi/version 0 47 0 compatible

FeignInt opened the pull request:
Reduces required meson version from 0.53.0 to 0.47.0

Inspired by:
https://github.com/dosbox-staging/dosbox-staging/issues/854#issuecomment-778892268  and https://github.com/mesonbuild/mt32emu/pull/2

I will happily squash the three commits (or drop either/both of the latter two)

dreamer commented on include/mt32emu/c_interface/meson.build:
```diff
@@ -1,3 +1,6 @@
-foreach file_name, src_file : mt32emu_c_header_map
-  configure_file(input : src_file, output : file_name, copy : true)
+foreach src_file : c_headers
+  file_name = 'string : @0@'.format(src_file).split('/')
```
What is this `string : ` syntax? Is it documented somewhere?

FeignInt replied:
https://mesonbuild.com/Syntax.html#string-formatting
and now that I re-read it, only need:
```
file_name = '@0@'.format(src_file).split('/')
```
I did initially try *src_file.split('/')* , but it wasn't happy, hence the format

dreamer commented on include/mt32emu/c_interface/meson.build:
```diff
@@ -1,3 +1,6 @@
+foreach src_file : c_headers
+  file_name = 'string : @0@'.format(src_file).split('/')
+  configure_file(input : src_file,
+                output : file_name[-1],
+                copy   : true)
```
One more space before `output` and `copy` is needed to keep the consistent style.

But I think this line is short and readable enough to fit into a single line :)

dreamer commented on include/mt32emu/meson.build:
```diff
@@ -2,8 +2,11 @@ configure_file(input : mt32emu_config_h_in,
 
-foreach file_name, src_file : mt32emu_header_map
-  configure_file(input : src_file, output : file_name, copy : true)
+foreach src_file : common_headers + cpp_headers
+  file_name = 'string : @0@'.format(src_file).split('/')
+  configure_file(input : src_file,
+                output : file_name[-1],
+                copy   : true)
```
*The same comments as for `c_interface` headers.*

dreamer commented on meson.build:
```diff
@@ -17,9 +17,12 @@ conf_data.set('libmt32emu_EXPORTS_TYPE',  mt32emu_exports_type)
                       cpp_args : ['-DMT32EMU_WITH_INTERNAL_RESAMPLER'])
 
-mt32emu_dep = declare_dependency(include_directories : include_directories('include'),
+mt32emu_dep = declare_dependency(include_directories: mt32emu_dep_includes,
```
Good change :)

Just please put a space between `include_directories` and `:` in here (to keep the style consistent with Meson default coding style).

dreamer commented on meson.build:
```diff
@@ -20,9 +20,9 @@ subdir('include/mt32emu') # public interface
-mt32emu_lib = library('mt32emu', mt32emu_sources,
-                      include_directories : mt32emu_lib_includes,
-                      cpp_args : ['-DMT32EMU_WITH_INTERNAL_RESAMPLER'])
+mt32emu_lib = static_library('mt32emu', mt32emu_sources,
```
Unfortunately, this change has bad consequence :(

Packagers and end-users want static library, yes - but developers want dynamic library to keep the builds faster (otherwise Meson can't use dynamic linking optimization).

So:
- Pro: this change makes it a bit easier to use for packagers and end-users
- Con: this adds some tiny amount of time to linking time, and I do linking many, many times a day

Therefore remove this change, please.

Do I understand correctly - this change was needed because `meson setup -Ddefault_library=static`  in dosbox-staging repo didn't work for 0.53.0, but does work for >= 0.54.0?

dreamer commented:
Overall I like these changes, but perhaps we can drop some of them if we agree that 0.53.0 shall be the minimum required version :)

eli-schwartz commented:
You could probably do entirely without dictionaries by building an array of arrays. Then for each elem in array, elem[0] and elem[1] are the input/output.

I would prefer this over the confusing string splitting...

FeignInt commented:
Thanks @eli-schwartz 

I have pushed a fixup,

dreamer reviewed (approved):
Looks great and :+1: for compatibility with 0.49.0 :)

I think it's time to squash into a single commit?

eli-schwartz commented:
Sounds good to me.

FeignInt committed:
Make compatible with meson v0.49.0 or greater

This commit avoids use of 0.53.0 feature:

  * {'Dictionary entry using non literal key'}

Also use include_directories() to avoid:

    ERROR:  Include directory to be added is not
            an include directory object.

Confirmed working with meson(1) versions:

    0.49.2 and 0.56.2
